[SwiftUI] Find-in-Page integration (macOS)

rr-codes opened the pull request:
#### c728b84463466fdcbc5215809c205d8fed589862
<pre>
[SwiftUI] Find-in-Page integration (macOS)
<a href="https://bugs.webkit.org/show_bug.cgi?id=285290">https://bugs.webkit.org/show_bug.cgi?id=285290</a>
<a href="https://rdar.apple.com/141503355">rdar://141503355</a>

Reviewed by Wenson Hsieh.

Use NSTextFinder with the backing WKWebView to integrate find-in-page capabilities in SwiftUI.

* Use autolayout instead of manual frame adjustment for `CocoaWebViewAdapter` so that the subviews can
easily adjust based on the presence or lack of find bar.

* Refactor CocoaWebViewAdapter a bit to be as platform agnostic as possible by introducing two helper protocols.

* Re-arrange and rename some files and types for better organization.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(backingWebView):
* Source/WebKit/UIProcess/API/Swift/WebView/CocoaWebViewAdapter.swift: Added.
(CocoaWebViewAdapter.isFindNavigatorVisible):
(CocoaWebViewAdapter.findInteraction):
(CocoaWebViewAdapter.isFindBarVisible):
(CocoaWebViewAdapter.findBarView):
(CocoaWebViewAdapter.findContext):
(CocoaWebViewAdapter.performFindPanelAction(_:)):
(CocoaWebViewAdapter.findBarDidBecomeVisible):
(CocoaWebViewAdapter.findBarDidBecomeHidden):
(CocoaWebViewAdapter.webViewConstraints):
(CocoaWebViewAdapter.findBarConstraints):
(CocoaWebViewAdapter.webViewHeightConstraint):
(CocoaWebViewAdapter.removeConstraints):
(CocoaWebViewAdapter.activateConstraints):
(CocoaWebViewAdapter.webView):
(CocoaWebViewAdapter.contentView):
(CocoaWebViewAdapter.findBarViewDidChangeHeight):
(CocoaWebViewAdapter.findInteraction(_:didBegin:)):
(CocoaWebViewAdapter.findInteraction(_:didEnd:)):
(CocoaWebViewAdapter.supportsTextReplacement):
* Source/WebKit/UIProcess/API/Swift/WebView/PlatformTextSearching.swift: Copied from Tools/SwiftBrowser/Source/SwiftBrowser.swift.
(PlatformTextSearching.isFindNavigatorVisible):
(PlatformTextSearching.findInteraction):
(PlatformFindInteraction.presentFindNavigator(_:)):
(PlatformFindInteraction.dismissFindNavigator):
(NSTextFinderAdapter.presentFindNavigator(_:)):
(NSTextFinderAdapter.dismissFindNavigator):
(UIFindInteractionAdapter.presentFindNavigator(_:)):
(UIFindInteractionAdapter.dismissFindNavigator):
* Source/WebKit/UIProcess/API/Swift/WebView/WebView.swift: Copied from Tools/SwiftBrowser/Source/SwiftBrowser.swift.
(WebView_v0.body):
* Source/WebKit/UIProcess/API/Swift/WebView/WebViewRepresentable.swift: Renamed from Source/WebKit/UIProcess/API/Swift/WebView.swift.
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewCoordinator.update(_:configuration:environment:)):
(WebViewCoordinator.updateFindInteraction(_:environment:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.body):

Canonical link: <a href="https://commits.webkit.org/288418@main">https://commits.webkit.org/288418@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e4322e4de98261108b2782e6fc2e7d324bd9027

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37463 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/34204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2881 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10733 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/88268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/34204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86223 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/2112 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75613 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/88268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/2017 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29814 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33241 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30522 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10448 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/7533 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/89637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10677 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71426 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/89637 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/16591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12858 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10402 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/10266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/13734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/12035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/API/Swift/WebView/CocoaWebViewAdapter.swift:
```diff
@@ -0,0 +1,252 @@
+    // MARK: PlatformTextSearching conformance
+
+#if os(macOS)
+    typealias Interaction = NSTextFinderAdapter
```
Should this be `typealias FindInteraction`? `Interaction` seems very generic, given the class name is `CocoaWebViewAdapter`, which is also a generic name.

whsieh commented on Source/WebKit/UIProcess/API/Swift/WebView/CocoaWebViewAdapter.swift:
```diff
@@ -0,0 +1,252 @@
+
+#if os(macOS)
+    typealias Interaction = NSTextFinderAdapter
+#else
```
`UIFindInteraction` is not supported on watchOS or tvOS. Does that need to be reflected here?

rr-codes replied:
nope, since ENABLE_SWIFTUI will not be true for those platforms

rr-codes committed:
[SwiftUI] Find-in-Page integration (macOS)
https://bugs.webkit.org/show_bug.cgi?id=285290
rdar://141503355

Reviewed by Wenson Hsieh.

Use NSTextFinder with the backing WKWebView to integrate find-in-page capabilities in SwiftUI.

* Use autolayout instead of manual frame adjustment for `CocoaWebViewAdapter` so that the subviews can
easily adjust based on the presence or lack of find bar.

* Refactor CocoaWebViewAdapter a bit to be as platform agnostic as possible by introducing two helper protocols.

* Re-arrange and rename some files and types for better organization.

* Source/WebKit/UIProcess/API/Swift/View+WebViewModifiers.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(backingWebView):
* Source/WebKit/UIProcess/API/Swift/WebView/CocoaWebViewAdapter.swift: Added.
(CocoaWebViewAdapter.isFindNavigatorVisible):
(CocoaWebViewAdapter.findInteraction):
(CocoaWebViewAdapter.isFindBarVisible):
(CocoaWebViewAdapter.findBarView):
(CocoaWebViewAdapter.findContext):
(CocoaWebViewAdapter.performFindPanelAction(_:)):
(CocoaWebViewAdapter.findBarDidBecomeVisible):
(CocoaWebViewAdapter.findBarDidBecomeHidden):
(CocoaWebViewAdapter.webViewConstraints):
(CocoaWebViewAdapter.findBarConstraints):
(CocoaWebViewAdapter.webViewHeightConstraint):
(CocoaWebViewAdapter.removeConstraints):
(CocoaWebViewAdapter.activateConstraints):
(CocoaWebViewAdapter.webView):
(CocoaWebViewAdapter.contentView):
(CocoaWebViewAdapter.findBarViewDidChangeHeight):
(CocoaWebViewAdapter.findInteraction(_:didBegin:)):
(CocoaWebViewAdapter.findInteraction(_:didEnd:)):
(CocoaWebViewAdapter.supportsTextReplacement):
* Source/WebKit/UIProcess/API/Swift/WebView/PlatformTextSearching.swift: Copied from Tools/SwiftBrowser/Source/SwiftBrowser.swift.
(PlatformTextSearching.isFindNavigatorVisible):
(PlatformTextSearching.findInteraction):
(PlatformFindInteraction.presentFindNavigator(_:)):
(PlatformFindInteraction.dismissFindNavigator):
(NSTextFinderAdapter.presentFindNavigator(_:)):
(NSTextFinderAdapter.dismissFindNavigator):
(UIFindInteractionAdapter.presentFindNavigator(_:)):
(UIFindInteractionAdapter.dismissFindNavigator):
* Source/WebKit/UIProcess/API/Swift/WebView/WebView.swift: Copied from Tools/SwiftBrowser/Source/SwiftBrowser.swift.
(WebView_v0.body):
* Source/WebKit/UIProcess/API/Swift/WebView/WebViewRepresentable.swift: Renamed from Source/WebKit/UIProcess/API/Swift/WebView.swift.
(WebViewRepresentable.makePlatformView(_:)):
(WebViewRepresentable.updatePlatformView(_:context:)):
(WebViewCoordinator.update(_:configuration:environment:)):
(WebViewCoordinator.updateFindInteraction(_:environment:)):
(WebViewRepresentable.makeUIView(_:)):
(WebViewRepresentable.updateUIView(_:context:)):
(WebViewRepresentable.makeNSView(_:)):
(WebViewRepresentable.updateNSView(_:context:)):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/SwiftBrowser/Source/SwiftBrowser.swift:
(SwiftBrowserApp.body):

Canonical link: https://commits.webkit.org/288418@main
